### PR TITLE
feat(graticule): coordinate grid overlay with edge labels

### PR DESCRIPTION
**Pull request by giswqs**
## Summary

Adds a **Graticule** plugin that draws a lat/long reference grid (meridians and parallels) with coordinate labels along the map edges, addressing the request in #766. The grid renders as native MapLibre `line` and `symbol` layers, so it is part of the GL canvas and is **captured automatically by the Print Layout export** (PNG/PDF), no separate print code needed.

Enable it from the **Plugins** menu; a small grid button is added to the map and opens the settings panel.

## Features

- **Spacing**: Auto (adapts to zoom with nice round intervals) or a fixed interval in degrees.
- **Line style**: color, width, opacity, and a dashed toggle.
- **Edge labels**: decimal-degree (`110°W`) or degrees/minutes/seconds (`122°30'00"W`) formatting; on the left + bottom edges or all four sides; with color and size controls.
- **Project persistence**: all settings round-trip through `.geolibre.json` via `getProjectState`/`applyProjectState`, validated and clamped on load so a hand-edited project cannot produce an invalid grid.

## Implementation notes

- New plugin `packages/plugins/src/plugins/maplibre-graticule.ts`, exported from the package `index.ts` and registered in `usePlugins.ts`, following the existing built-in plugin pattern.
- Grid geometry is recomputed on `moveend`; the layers are rebuilt on basemap change (since `setStyle` drops user sources/layers).
- Labels reuse a font the active basemap already ships (preferring an upright regular face), so the glyphs load reliably across basemaps rather than depending on a hard-coded font name.

## Scope

This first pass covers the geographic (lat/long) grid type. Projected-CRS grids are a natural follow-up. On the globe projection the grid lines render correctly; edge labels are placed at the viewport bounds, which is best suited to the flat/Mercator view.

## Testing

- New unit tests (`tests/graticule.test.ts`, 10 cases) cover settings normalization/clamping and decimal-degree + DMS label formatting, including longitude wrap-around.
- Full frontend suite passes (1454 tests), typecheck passes, eslint and the build pre-commit hook pass.
- Verified in the real app with Playwright: the grid and edge labels render in both **light and dark** themes (confirmed labels such as `110°W` on the bottom edge and `50°N` on the left edge), the settings panel and on-map button work, and toggling the plugin off cleanly removes the layers, control, and panel with no console errors.

## Release Notes

* **New Features**
  * Added a Graticule map control to display latitude/longitude grid lines with optional coordinate labels.
  * Introduced a Graticule settings panel (auto/fixed spacing, line styling, label format/placement) with new English localization.
  * Updated controls/menus so Graticule appears under map controls (and prints/layouts can account for label presence).

* **Bug Fixes**
  * Improved Graticule settings validation, clamping, and normalization with safer defaults.

* **Tests**
  * Added coverage for settings normalization and label formatting (dd/dms, wrapping, and edge cases).

### CODE REVIEW

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 1]
```diff
@@ -0,0 +1,684 @@
...
+    labelEdges: v.labelEdges === "all" ? "all" : "left-bottom",
+    labelColor: isHexColor(v.labelColor) ? v.labelColor : d.labelColor,
+    labelSize: clampNumber(v.labelSize, 6, 28, d.labelSize),
+  };
+}
+
+function isDefaultSettings(value: GraticuleSettings): boolean {
+  return (
+    JSON.stringify(value) === JSON.stringify(DEFAULT_GRATICULE_SETTINGS)
+  );
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts]: Done in a036641 — replaced with a field-by-field comparison against the defaults.

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 2]
```diff
@@ -0,0 +1,684 @@
...
+// Geometry generation
+// ---------------------------------------------------------------------------
+
+/** Pick an auto interval that draws roughly 4-12 meridians across the view. */
+function autoStep(bounds: LngLatBounds): number {
+  const lonSpan = Math.abs(bounds.getEast() - bounds.getWest()) || 0.001;
+  for (const step of NICE_STEPS) {
+    if (lonSpan / step >= 4) return step;
+  }
+  return NICE_STEPS[NICE_STEPS.length - 1];
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts]: Done in a036641 — added `unwrappedLongitudeRange` so antimeridian-crossing views (east < west) iterate an increasing range reused by `autoStep`, meridians, parallels, and edge labels.

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 3]
```diff
@@ -0,0 +1,684 @@
...
+ * the grid is part of the GL canvas and is captured automatically by the Print
+ * Layout export (PNG/PDF). All settings round-trip through the project file.
+ */
+
+export const GRATICULE_PLUGIN_ID = "maplibre-gl-graticule";
+
+const LINE_SOURCE_ID = "geolibre-graticule-lines";
+const LABEL_SOURCE_ID = "geolibre-graticule-labels";
+const LINE_LAYER_ID = "geolibre-graticule-lines";
+const LABEL_LAYER_ID = "geolibre-graticule-labels";
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts]: Done in a036641 — source and layer ids now use distinct `-source`/`-layer` suffixes.

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 4]
```diff
@@ -0,0 +1,684 @@
...
+
+  select(
+    "Spacing",
+    [
+      { value: "auto", label: "Auto (by zoom)" },
+      { value: "fixed", label: "Fixed interval" },
+    ],
+    () => settings.spacingMode,
+    (v) => setGraticuleSettings({ spacingMode: v as GraticuleSettings["spacingMode"] }),
+  );
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts:718]: Done in a036641 — all panel/control strings now go through i18n via a `setGraticuleLabels` setter populated by `t()` in `TopToolbar`, backed by a new `graticule` catalog in `en.json` (the same plain-DOM label-injection pattern as `maplibre-reverse-geocode`).

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 5]
```diff
@@ -0,0 +1,684 @@
...
+  }
+  ensureLayers(activeMap);
+  const geometry = buildGeometry(activeMap);
+  (activeMap.getSource(LINE_SOURCE_ID) as GeoJSONSource | undefined)?.setData(
+    geometry.lines,
+  );
+  (activeMap.getSource(LABEL_SOURCE_ID) as GeoJSONSource | undefined)?.setData(
+    geometry.labels,
+  );
+  applyStyleProps(activeMap);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts]: Done in a036641 — split into `refreshGeometry` (geometry only, used on pan/zoom) and `update` (also re-applies paint/layout props, used on settings/basemap change).

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 6]
```diff
@@ -0,0 +1,684 @@
...
+    "top",
+    ["literal", [0, 0.5]],
+    "left",
+    ["literal", [0.5, 0]],
+    "right",
+    ["literal", [-0.5, 0]],
+    ["literal", [0, 0]],
+  ]);
+  activeMap.setPaintProperty(LABEL_LAYER_ID, "text-color", settings.labelColor);
+  activeMap.setPaintProperty(LABEL_LAYER_ID, "text-halo-color", "#ffffff");
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts]: Done in a036641 — the halo colour is now derived from the label colour's luminance (dark halo for light text, white for dark), so labels stay legible on dark basemaps.

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 7]
```diff
@@ -0,0 +1,684 @@
...
+    set: (value: string) => void,
+  ): void => {
+    const el = document.createElement("input");
+    el.type = "color";
+    el.value = get();
+    el.addEventListener("input", () => set(el.value));
+    controls.push(() => {
+      el.value = get();
+    });
+    addRow(labelText, el);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts:692]: Done in a036641 — `normalizeHexColor` now canonicalizes to lowercase `#rrggbb` (expanding `#rgb`) and rejects 5/7-digit and `#rrggbbaa` forms the colour input cannot display; updated the test expectation to `#00aa00`.

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 8]
```diff
@@ -0,0 +1,684 @@
...
+      { value: "fixed", label: "Fixed interval" },
+    ],
+    () => settings.spacingMode,
+    (v) => setGraticuleSettings({ spacingMode: v as GraticuleSettings["spacingMode"] }),
+  );
+  number(
+    "Interval (°)",
+    { min: 0.001, max: 45, step: 0.5 },
+    () => settings.spacingDegrees,
+    (v) => setGraticuleSettings({ spacingDegrees: v }),
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts:725]: Done in a036641 — the interval input now uses step=0.001 so default/clamped values (10, 0.25, ...) are valid for the native number input.

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 9]
```diff
@@ -0,0 +1,684 @@
...
+    unregisterPanel =
+      app.registerRightPanel?.({
+        id: PANEL_ID,
+        title: "Graticule",
+        dock: "right-of-style",
+        render: (container) => renderPanel(container),
+      }) ?? null;
+
+    control = new GraticuleControl();
+    app.addMapControl(control, "top-right");
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts]: Done in a036641 — `addMapControl`'s return value is now checked; the control is nulled when the add is refused so `deactivate` cannot remove a dangling reference.

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 10]
```diff
@@ -0,0 +1,684 @@
...
+  activeMap.setPaintProperty(LABEL_LAYER_ID, "text-halo-color", "#ffffff");
+  activeMap.setPaintProperty(LABEL_LAYER_ID, "text-halo-width", 1.2);
+}
+
+/** Recompute geometry and re-apply styling. Safe to call repeatedly. */
+function update(): void {
+  if (!map) return;
+  const activeMap = map;
+  if (!activeMap.isStyleLoaded()) {
+    activeMap.once("idle", () => update());
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts]: Done in a036641 — added an `idlePending` guard so at most one `idle` listener is queued instead of one per `update()` call during style load.

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 11]
```diff
@@ -0,0 +1,831 @@
...
+    labels.labelSize,
+    { min: 6, max: 28, step: 1 },
+    () => settings.labelSize,
+    (v) => setGraticuleSettings({ labelSize: v }),
+  );
+
+  syncPanel = () => {
+    for (const sync of controls) sync();
+  };
+  return () => {
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts]: Done in a087364 — `renderPanel`'s cleanup now keys on container identity (`panelContainer === container`) instead of truthiness, so a second render landing before the first cleanup fires can no longer null the newer render's `syncPanel`.

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 12]
```diff
@@ -0,0 +1,831 @@
...
+
+class GraticuleControl implements IControl {
+  private container: HTMLElement | null = null;
+
+  onAdd(): HTMLElement {
+    const container = document.createElement("div");
+    container.className = "maplibregl-ctrl maplibregl-ctrl-group";
+    const button = document.createElement("button");
+    button.type = "button";
+    button.title = labels.controlTitle;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts]: Done in a087364 — `GraticuleControl` now stores its button and exposes `updateLabels()`, and `setGraticuleLabels` calls it plus rebuilds the open panel body (extracted into `buildPanelBody`) so the tooltip/aria-label and panel controls follow a mid-session language change. One remnant: the panel's header title is owned by the host (passed once to `registerRightPanel`), so it refreshes when the panel is reopened rather than live; happy to wire a host-side title updater in a follow-up if you'd prefer it fully live.

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 13]
```diff
@@ -0,0 +1,863 @@
...
+}
+
+/** Pick an auto interval that draws roughly 4-12 meridians across the view. */
+function autoStep(lonSpan: number): number {
+  const span = Math.abs(lonSpan) || 0.001;
+  for (const step of NICE_STEPS) {
+    if (span / step >= 4) return step;
+  }
+  return NICE_STEPS[NICE_STEPS.length - 1];
+}
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts]: Done in 36eb461 — `autoStep` now takes both spans and uses `max(lonSpan, latSpan)`, so a tall polar view no longer draws far more parallels than meridians.

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 14]
```diff
@@ -0,0 +1,863 @@
...
+      control = null;
+    }
+    unregisterPanel?.();
+    unregisterPanel = null;
+    syncPanel = null;
+    panelContainer = null;
+    if (map) teardownLayers(map);
+    map = null;
+    appRef = null;
+  },
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts:902]: Done in 36eb461 — `deactivate` now resets `idlePending = false`, so a re-activation during style load queues its own idle draw instead of waiting on the previous run's stale listener.

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 15]
```diff
@@ -0,0 +1,863 @@
...
+/**
+ * Replace the user-facing strings (the host calls this with translations on
+ * every language change). Pushes the new strings into the live control tooltip
+ * and, if the settings panel is open, rebuilds its body so labels stay current.
+ */
+export function setGraticuleLabels(next: Partial<GraticuleLabels>): void {
+  labels = { ...labels, ...next };
+  control?.updateLabels();
+  if (panelContainer) buildPanelBody(panelContainer);
+}
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts:106]: Done in 36eb461 — added a doc comment on `setGraticuleLabels` noting the header title is fixed at `registerRightPanel` time (no host API to update it) and so only re-localizes on reopen, unlike the body and control tooltip which now update live.

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 16]
```diff
@@ -0,0 +1,863 @@
...
+  for (let lat = firstLat; lat <= north && count < maxLines; lat += step) {
+    lineFeatures.push({
+      type: "Feature",
+      properties: {},
+      geometry: { type: "LineString", coordinates: densifyLine(lat, west, east, "lat") },
+    });
+    if (settings.showLabels) {
+      labelFeatures.push(labelFeature(west, lat, formatLat(lat, step, settings.labelFormat), "left"));
+      if (showAllEdges) {
+        labelFeatures.push(labelFeature(east, lat, formatLat(lat, step, settings.labelFormat), "right"));
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts:342]: Done in 36eb461 — added a comment explaining edge-label longitudes may exceed [-180, 180] for antimeridian-crossing views (correct for MapLibre's continuous world; only a raw-FeatureCollection consumer would need to wrap).

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 17]
```diff
@@ -0,0 +1,863 @@
...
+/**
+ * Decimal places to show for a given interval, derived from the step's own
+ * precision so labels are neither lossy (e.g. 1.25 shown as "1.3" for a 0.25
+ * interval) nor needlessly long. Capped at 4 places.
+ */
+function decimalsForStep(step: number): number {
+  const text = String(step);
+  const dot = text.indexOf(".");
+  if (dot === -1) return 0;
+  return Math.min(4, text.length - dot - 1);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts:223]: Done in 36eb461 — `decimalsForStep` now guards against scientific-notation strings (returns max precision), making the >=0.001 invariant explicit even though the clamp/NICE_STEPS floor keeps it unreachable in practice.

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 18]
```diff
@@ -0,0 +1,883 @@
...
+    lineColor: normalizeHexColor(v.lineColor) ?? d.lineColor,
+    lineWidth: clampNumber(v.lineWidth, 0.1, 6, d.lineWidth),
+    lineOpacity: clampNumber(v.lineOpacity, 0, 1, d.lineOpacity),
+    lineDashed: typeof v.lineDashed === "boolean" ? v.lineDashed : d.lineDashed,
+    showLabels: typeof v.showLabels === "boolean" ? v.showLabels : d.showLabels,
+    labelFormat: v.labelFormat === "dms" ? "dms" : "dd",
+    labelEdges: v.labelEdges === "all" ? "all" : "left-bottom",
+    labelColor: normalizeHexColor(v.labelColor) ?? d.labelColor,
+    labelSize: clampNumber(v.labelSize, 6, 28, d.labelSize),
+  };
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts:811]: Done in 4d4ce9f — addressed via the font cache (1a): `pickTextFont` is memoized so the colour picker's continuous `input` events no longer trigger an O(n_layers) scan; each event now only rebuilds the (bounded) grid geometry and sets paint/layout props, so live preview stays responsive without a debounce.

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 19]
```diff
@@ -0,0 +1,883 @@
...
+    return container;
+  }
+
+  /** Refresh the tooltip/aria-label so they follow a language change. */
+  updateLabels(): void {
+    if (!this.button) return;
+    this.button.title = labels.controlTitle;
+    this.button.setAttribute("aria-label", labels.controlTitle);
+  }
+
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts:579]: Done in 4d4ce9f — `pickTextFont` now caches its result (a module-level `cachedTextFont`, invalidated on basemap change and deactivate), so `applyStyleProps` no longer rescans every style layer on each colour-picker input event.

File: packages/plugins/src/plugins/maplibre-graticule.ts [hunk 20]
```diff
@@ -0,0 +1,883 @@
...
+    if (map) teardownLayers(map);
+    map = null;
+    appRef = null;
+  },
+  getProjectState: () => (isDefaultSettings(settings) ? undefined : { ...settings }),
+  applyProjectState: (_app: GeoLibreAppAPI, state: unknown) => {
+    settings = normalizeGraticuleSettings(state);
+    update();
+    syncPanel?.();
+  },
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/maplibre-graticule.ts:912]: Done in 4d4ce9f — `applyProjectState` now returns early (no `update()`/`syncPanel`) when the normalized incoming state already equals the in-memory settings, via a shared `settingsEqual` helper, matching the maplibre-effects pattern.
